### ISSUE TITLE
Protocol documentation?

### ISSUE THREAD

**Issue by christianbundy** [hyperswarm/hyperswarm#60]
Hi! I'm looking for docs to implement this in other languages. My understanding is that it's DHT + MDNS, but I'm not sure how straightforward it is. Are there any resources y'all could recommend for making Hyperswarm work outside of Node.js? Thanks! :star:

**Comment by mafintosh** [comment]
I’m in the middle of documenting it right now :) We’ll be putting up a website as well. Which languages are you porting it to?

**Comment by christianbundy** [comment]
Great! I'm not actively working on porting to any language, but after investing heaps of time learning the SSB stack I'm hesitant to adopt a reference implementation before I can understand how hard it'd be to implement from scratch. Hyperswarm is lovely to work with because it Just Works, but I don't actually know what's Just Working and so I figured I'd reach out and start a chat.

Thanks for all of your work on this project!

**Comment by mafintosh** [comment]
Makes tons of sense. The DHT protocol is not too complicated (at least compared to the rest of the Hyper stack).

Since there’s growing interest I’ll try to just get the content up and style it later to be all nice and stuff :)

**Comment by christianbundy** [comment]
Sounds great, looking forward to checking it at your convenience.

**Comment by christianbundy** [comment]
Hi, I'm checking back in on this -- are there any resources I should check out, or are the protocol docs still WIP?

**Comment by decentral1se** [comment]
Yes, just wanted to chime in that this would be utterly fantastic (having the python implementation in mind...) :sun_with_face:

**Comment by amoallim15** [comment]
I'm considering to port this specific implementation to python :D Keep up the great work!

**Comment by decentral1se** [comment]
Hey @amoallim15 if you get started and want to chat, I've been working on porting the hyper stack to python over at https://hyperpy.decentral1.se so would happily try to help out with your effort! Would be cool to see more python stuff in this space.

**Comment by smoyer64** [comment]
@mafintosh 

I was working on a Go implementation last year and have just come back to working on the project.  The changes from v2.x.x to v3.x.x eliminate the use of protobuf and so the above referenced documentation is more important.  What's the status of the protocol documentation?  Would you want help with it?  Or proof-reading it?  (I'm a long-time protocol engineer and worked on the DOCSIS 3.0 and 3.1 specifications so I've got an eye tuned towards separating the protocol specifics from the implementation.

**Comment by eshork** [comment]
Hi! 👋
I just finished a full wire-compatible Rust port of the entire Hyperswarm + HyperDHT stack (including the UDX transport layer).
While building it I ended up writing a detailed protocol spec for UDX (packet format, flags, BBR congestion control, SACK, stream multiplexing, relay, RTT estimation, MTU discovery, etc.):
→ https://github.com/Rightbracket/peeroxide/blob/main/libudx/PROTOCOL.md
It’s reverse-engineered from the canonical C reference (holepunchto/libudx) with byte-level layouts and notes on the differences/simplifications in my implementation. Should be useful if you’re still working on the official docs — or for the folks working on the Go port.
The repo also includes golden test fixtures and live cross-language interop tests against the Node.js versions if that’s helpful.
